### DISCUSSION TITLE
New blog post: Tailwind vs semantic CSS

### DISCUSSION THREAD

**Discussion by tipiirai** [nuejs/nue#86] [Announcements]
Check out:

https://nuejs.org/blog/tailwind-vs-semantic-css/

**Comment by felix-bohlin** [comment]
I thought it was very interesting when you pointed out the future technical debt Tailwind could become. Haven't really considered that viewpoint.

**Reply by tipiirai** [reply]
That is certainly a risk. We might look back at the odd-looking, highly verbose TW syntax after a few years and think... oh, we did that?! There are way fewer risks by sticking closer to standards and using CSS as intended.

**Reply by felix-bohlin** [reply]
Well the problem as I see it is to get people to care about CSS. Most devs just doesn't seem interested, but maybe it has to come to this to get people to change their minds, I'm not sure.

**Reply by tipiirai** [reply]
Agreed. That's a challenge, for sure. One strategy is to provide high quality templates with stellar design.

**Comment by siduck** [comment]
i agree with the tailwind points. Although I use unocss which seems better to me, especially some of its features like

1) Attributify plugin & valueless attributes

```html
<div bg='red' border='1 solid blue'> 
```
instead of tailwind's className='border-1 border-solid border-blue`

it gets cleaner with valueless attributes, like 

<div bg-red p5>

Another thing i like about this is that the responsive related classes are easy to use, like

`<div   lg="bg-red p5" p10> ` or `<div p="lg:5 10">`

 2) Shortcuts , along with tagify plugin
 
 
![image](https://github.com/nuejs/nue/assets/59060246/487e1425-c7dd-47ab-a605-66044ddbb1ac)

here shortcuts could be used as classname / tag name, like

 `<card><card/>`  `<full-overlay> <full-overlay/>` would be   valid and still work ! 
 
  you can even compile the classes into 1 
  
  
![image](https://github.com/nuejs/nue/assets/59060246/e834450c-7a97-4207-af8e-63a8191a2f91)

Imo unocss can be a lot more cleaner than tailwind,  I like styling in the markup cuz its easy to make responsive quick

**Reply by tipiirai** [reply]
Personally I divide CSS development in two camps

1: CSS-IN-JS 
2. External CSS

The JS camp favors tight coupling (the styling and the component belong together) and the CSS camp respects the separation of concerns principle. These two make a whole different development model. UnoCSS belongs to the JS camp.

**Reply by eweren** [reply]
I guess UnoCSS can be seen in both camps.
You can either use it as tailwind alternative, tightly coupling the styles to the markup or you use it to create standard css classes decoupled from markup faster as before.

Instead of standard tailwind
```
<div class="bg-red text-blue w-40 h-40 lg:w-60 lg:h-60" />
```

You can make it simpler by using `--at-apply` within your css which then takes the basic rules of tailwind/unocss and compiles them into plain css. Like so:

```
<div class="colored-container" />

// with

.colored-container {
  --at-apply: bg-red text-blue w-40 h-40 lg:w-60 lg:h-60;
}

// which compiles to the equivalent rules in one plain css class with media queries.
```

The drawback is that it has to get compiled - even though UnoCSS is blazing fast. But you get faster dev time (imo) especially when it gets to writing media queries or pseudo status code. Just add your rule to --at-apply instead of writing 3 or more lines just to get your screen sizes right.

I see your point, but I guess it is not black and white. Technologies like UnoCSS try to bridge the gap between writing the whole CSS on your own and using bloated tailwind components while relying on standards like css variables.

**Reply by siduck** [reply]
@eweren i think that can be done with tailwind too, but unocss aims to be minimal. I like its attributify plugin and with its tagify plugin i could just maintain a list of shortcuts and use them as custom tag names

**Reply by tipiirai** [reply]
So Uno is similar to Daisy UI or Tailwind @apply for those who attempt to get benefits of both utility classes and semantic naming? Perhaps smaller in scope than Daisy?

**Reply by siduck** [reply]
@tipiirai unocss is an alternative to tailwind, its just that you get more powers that lead to cleaner syntax via its plugins, unlike with tailwind.

**Reply by tipiirai** [reply]
More powerful and cleaner syntax — that's all we need! Uno CSS sounds like semantic CSS then.

**Reply by eweren** [reply]
@siduck you are right. The `--at-apply` UnoCSS I mentioned can be done in a similar way (`@apply` which will also work work with uno) with tailwind.
While you can use `@apply` in plain css files it will not work in svelte files or other framework markup file styles-blocks because it is tailwind specific syntax and will not work from within those style blocks. UnoCSS instead is using the cssvar systax (as fallback) and therefore it works in those files too. Maybe it was because of that why I forgot that tailwind got this functionality too 😅

**Reply by siduck** [reply]
@eweren @apply works in css files for svelte for me if i load the css file in the main layout.svelte file! 

with unocss we get tailwind's functionalities as well as some goodies like class attributes & tagify plugin! you can compile all of its classes into one single class too but youve to mention `class=":uno: text-lg` but its too much of work to do it for every class lol

@tipiirai  it'd be great if nue could allow usage of custom tag names which are named after class names. Unocss does that with unocss shortcuts like

if a shortcut's like `overlay: "h-full absolute inset-0 bg-white"` , it could be used as `<overlay></overlay>` , But the caveat for this is that if in future if html introduces same tag names then it'd be a problem! unocss could just compile those classnames into `<div class='overlay'>` but it doesnt

also its attributes can be used with the custom tags, like

`<overlay bg='blue lg:red' xl="p5>`

**Comment by inwardmovement** [comment]
Maybe someday we'll see something that allows us to use Tailwind in dev, but that translates to "optimized HTML/CSS" at build time, to get around the blotted HTML/CSS argument of the blog post... best of both worlds (if one doesn't want to deal with naming/organizing CSS).

**Reply by tipiirai** [reply]
Nue will support Tailwind in the future in such a way that the production CSS is stripped out of unused class names. Of course, Tailwind can never beat well-structured semantic CSS regarding file size, even with optimization. I expect that Tailwind developers have optimized the Spotlight template being compared on the article.

**Reply by felix-bohlin** [reply]
Doesn't the Tailwind JIT-compiler already do that? Or is it on a per project basis and not per page/component basis?

**Reply by inwardmovement** [reply]
> Doesn't the Tailwind JIT-compiler already do that? Or is it on a per project basis and not per page/component basis?

No it doesn't reorganise all your CSS and rename your classes like it's showed in the article.

**Reply by felix-bohlin** [reply]
Sorry, it was in response to:
> Nue will support Tailwind in the future in such a way that the production CSS is stripped out of unused class names.

But maybe I have it the wrong way around here too.

**Reply by inwardmovement** [reply]
> Sorry, it was in response to:
> > Nue will support Tailwind in the future in such a way that the production CSS is stripped out of unused class names.

Oh sorry! Yes the last version of Tailwind only includes the used classes by default.

**Comment by felix-bohlin** [comment]
Maybe you've explained that somewhere, but what is your vision for Nue CSS? We have [Open Props](https://open-props.style/) all the way to UnoCSS, but I'm not sure really what problems you're aiming to solve apart from pushing it closer to standards.

**Reply by siduck** [reply]
@felix-bohlin many like tailwind because it lets you do styling in the markup file itself, therefore reducing dev time on opening other css files by a lot.  I'd still go for unocss as it lets you do what tailwind does but with more plugins as goodies making a clean experience unlike tailwind

**Reply by felix-bohlin** [reply]
Yes for sure, I agree - UnoCSS does a lot of things right, but since the [roadmap](https://nuejs.org/blog/perfect-web-framework/#product-roadmap) includes a CSS design system I'm curious how that will differ from what exists today. Maybe I understood it wrong somehow, but either way I'm curious for some clarification.

**Reply by tipiirai** [reply]
Roughly: Nue CSS is a design system for both multi- and single-page applications. It lets you take advantage of both CSS-in-JS and external/cascaded CSS. ie. take advantage of different cascade layers (root > area > page) without the issues of the global namespace. A little abstract at this point, but it will all make sense once released.
